### PR TITLE
Fix/csv export attachments

### PR DESCRIPTION
**Pull request by hjJunior**
### Fixed
- Imports notes of stories

### Added
- Imports/exports of stories 
  - Documents
  - Tasks

### CODE REVIEW

File: app/models/task.rb [hunk 1]
```diff
@@ -6,4 +6,18 @@ class Task < ApplicationRecord
...
+    [name, status]
+  end
+
+  private
+
+  def status
+    if done
+      'completed'
+    else
+      'not_completed'
```
Comment (matheusazzi) [inline app/models/task.rb]: @hjJunior It looks like this branch of code isn't tested. I think we should add a spec for when `done` is false.

Comment (hjJunior) [reply app/models/task.rb]: I'll work to add this spec, thanks

File: spec/models/note_spec.rb [hunk 2]
```diff
@@ -24,14 +24,14 @@
...
-  describe '#to_s' do
+  describe '#to_csv' do
     before do
       subject.note = 'Test note'
       subject.created_at = 'Nov 3, 2011'
       allow(user).to receive_messages(name: 'user')
     end
 
-    its(:to_s) { should == 'Test note (user - Nov 03, 2011)' }
+    its(:to_csv) { should == 'Test note (user - Nov 03, 2011)' }
```
Comment (brunohkbx) [inline spec/models/note_spec.rb]: What do you think about rewrite this spec using the it/expect syntax?
We should be consistent.

File: spec/models/task_spec.rb [hunk 3]
```diff
@@ -22,6 +22,14 @@
     end
   end
 
+  describe '#to_csv' do
+    subject(:task) { build :task, story: story, name: 'task_test', done: true }
```
Comment (brunohkbx) [inline spec/models/task_spec.rb]:
```suggestion
    let(:task) { build(:task, story: story, name: 'task_test', done: true) }
    
    subject { task.to_csv }
    
    it { is_expected.to contain_exactly(task.name, 'completed') }
```

File: config/initializers/csv_renderer.rb [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
+      csv << story.to_csv(number_of_extra_columns)
     end
   end
 
   send_data csv_string, type: Mime::CSV, filename: filename
 
 end
+
+def extra_headers(number_of_extra_columns)
+  [ Array.new(number_of_extra_columns[:notes], "Note"),
```
Comment (marclerodrigues) [inline config/initializers/csv_renderer.rb]: Remove space around `[` and `]`

File: spec/models/task_spec.rb [hunk 5]
```diff
@@ -22,6 +22,14 @@
...
+    
+    it 'return a array with name and status from task' do
```
Comment (marclerodrigues) [inline spec/models/task_spec.rb]: `it 'returns...`

File: app/models/task.rb [hunk 6]
```diff
@@ -6,4 +6,18 @@ class Task < ApplicationRecord
...
 
   delegate :readonly?, to: :story
+
+  def to_csv
+    [name, status]
+  end
+
+  private
+
+  def status
```
Comment (brunohkbx) [inline app/models/task.rb]:
```suggestion
  def status
    done ? 'completed' : 'not_completed'
  end
```

File: spec/models/note_spec.rb [hunk 7]
```diff
@@ -24,14 +24,16 @@
...
-  describe '#to_s' do
+  describe '#to_csv' do
     before do
       subject.note = 'Test note'
       subject.created_at = 'Nov 3, 2011'
       allow(user).to receive_messages(name: 'user')
     end
 
-    its(:to_s) { should == 'Test note (user - Nov 03, 2011)' }
+    it 'should have the format "task_name (user_name - date)"' do
```
Comment (brunohkbx) [inline spec/models/note_spec.rb]: Could you remove the should from this spec description?
http://www.betterspecs.org/#should

Comment (hjJunior) [reply spec/models/note_spec.rb]: What do think about: `gets formated task "task_name (user_name - date)"` is It better?

Comment (brunohkbx) [reply spec/models/note_spec.rb]: @hjJunior you didn't check the case when the user aren't present, so you need to create another spec to verify.

Check it out and let me know if this works.

```rb
context 'When Note has a user' do
  let(:note) do
    build_stubbed(
      :note,
      user: user,
      note: 'Test note',
      created_at: 'Nov 3, 2011'
    )
  end

  before { allow(user).to receive_messages(name: 'user') }

  subject { note.to_csv }

  it { is_expected.to eq('Test note (user - Nov 03, 2011)') }
end

context 'When Note does not have a user' do
  let(:user_name) { I18n.t('author unknown') }

  let(:note) do
    build_stubbed(:note, note: 'Test note', created_at: 'Nov 3, 2011')
  end

  subject { note.to_csv }

  it { is_expected.to eq('Test note (#{user_name} - Nov 03, 2011)') }
end
```

File: yarn.lock [hunk 8]
```diff
@@ -5,52 +5,62 @@
 abbrev@1:
   version "1.1.1"
   resolved "https://registry.yarnpkg.com/abbrev/-/abbrev-1.1.1.tgz#f8f2c887ad10bf67f634f005b6987fed3179aac8"
+  integrity sha512-nne9/IiQ/hzIhY6pdDnbBtz7DjPTKrY00P/zvPSm5pOFkl6xuGrGnXn/VtTNNfNtAfZ9/1RtehkszU9qcTii0Q==
```
Comment (brunohkbx) [inline yarn.lock]: @talyssonoc are these changes ok?

File: Gemfile.lock [hunk 9]
```diff
@@ -614,4 +614,4 @@ RUBY VERSION
    ruby 2.3.1p112
 
 BUNDLED WITH
-   1.16.4
+   1.16.6
```
Comment (talyssonoc) [inline Gemfile.lock]: I think we shouldn't update this since it's not part of the task.

File: config/initializers/attachinary.rb [hunk 10]
```diff
@@ -5,6 +5,10 @@ module FileMixin
     def as_json(options = {})
       super(only: [:id, :public_id, :format, :version, :resource_type], methods: [:path])
     end
+
+    def to_csv 
+      attributes.except("id","attachinariable_id","created_at", "updated_at")
```
Comment (talyssonoc) [inline config/initializers/attachinary.rb]: Can we use symbols here?

File: config/initializers/csv_renderer.rb [hunk 11]
```diff
@@ -1,27 +1,29 @@
...
+      csv << story.to_csv(number_of_extra_columns)
     end
   end
 
   send_data csv_string, type: Mime::CSV, filename: filename
 
 end
+
+def extra_headers(number_of_extra_columns)
+  [Array.new(number_of_extra_columns[:notes], "Note"),
```
Comment (andrerpbts) [inline config/initializers/csv_renderer.rb]: just a small 💅 

```ruby
def extra_headers(number_of_extra_columns)
  [
    Array.new(number_of_extra_columns[:notes], "Note"),
    Array.new(number_of_extra_columns[:documents], "Document"),
    Array.new(number_of_extra_columns[:tasks], ["Task", "Task Status"]).flatten
  ].flatten
end
```

File: config/initializers/csv_renderer.rb [hunk 12]
```diff
@@ -1,27 +1,29 @@
 # Currently only used for rendering collections of stories, so make an
 # assumption that that is what we get passed.
 ActionController::Renderers.add :csv do |stories, options|
-
+  number_of_extra_columns = {}
+  number_of_extra_columns[:notes] = stories.map{|story| story.notes.length}.max
```
Comment (andrerpbts) [inline config/initializers/csv_renderer.rb]: Rubocop isn't set for this project?

It should tell you to put a single space before/after curly brackets...

File: CHANGELOG.md [hunk 13]
```diff
@@ -2,9 +2,15 @@
...
 ### Fixed
-- Fix pusher missing api key
+- Imports notes of stories
 - Fix project delete when tasks or notes are readonly
 
+### Added
+- Imports/exports of stories 
+  - Documents
+  - Tasks
+- Fix pusher missing api key
```
Comment (talyssonoc) [inline CHANGELOG.md]: These two fixes should be in the `fixed` section.

File: Gemfile [hunk 14]
```diff
@@ -14,7 +14,7 @@ gem 'api-pagination'
 gem 'attachinary'
 gem 'autoprefixer-rails'
 gem 'bootstrap-sass', '~> 3.3.5'
-gem 'central-support', github: 'Codeminer42/cm42-central-support', require: 'central/support'
+gem 'central-support', github: 'Codeminer42/cm42-central-support', require: 'central/support', branch: 'fix/csv-export-attachments'
```
Comment (talyssonoc) [inline Gemfile]: I just merged the PR from the `cm42-central-support` repository, we can remove the `branch` option now 😄

**Pull request merged**